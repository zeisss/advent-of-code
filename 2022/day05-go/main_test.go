package main

import (
	"fmt"
	"os"
	"testing"

	approvals "github.com/approvals/go-approval-tests"
	"github.com/stretchr/testify/require"
)

func TestApplyAction(t *testing.T) {
	for i := 1; i <= 3; i++ {
		t.Run(fmt.Sprintf("amount=%d", i), func(t *testing.T) {
			out := StackMover9000.applyAction(
				[][]string{
					{"a", "b", "c"},
					{},
				},
				Action{Src: 1, Dst: 2, Amount: i},
			)

			require.Len(t, out[0], 3-i)
			require.Len(t, out[1], i)
		})
	}
}

func TestGetTopContainerNames(t *testing.T) {
	out := [][]string{
		{"a", "b", "c"},
		{"d"},
	}
	s := GetTopContainerNames(out)
	require.Equal(t, "cd", s)
}

func TestParseData(t *testing.T) {
	input := `    [D]    
[N] [C]    
[Z] [M] [P]
 1   2   3 

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2`
	plan, err := ParseData([]byte(input))
	require.NoError(t, err)

	approvals.VerifyJSONStruct(t, plan)
}

func TestParseData_Actions(t *testing.T) {
	for actions := 1; actions <= 4; actions++ {
		t.Run(fmt.Sprintf("processNextAction() %dx", actions), func(t *testing.T) {
			input := `    [D]    
[N] [C]    
[Z] [M] [P]
1   2   3 

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2`
			plan, err := ParseData([]byte(input))
			require.NoError(t, err)

			for i := 0; i < actions; i++ {
				StackMover9000.processNextAction(&plan)
			}
			approvals.VerifyJSONStruct(t, plan)

		})
	}
}

func TestMain(m *testing.M) {
	approvals.UseFolder("testdata")

	os.Exit(m.Run())
}
