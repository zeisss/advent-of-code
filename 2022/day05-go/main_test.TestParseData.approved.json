{
  "Stacks": [
    [
      "Z",
      "N"
    ],
    [
      "M",
      "C",
      "D"
    ],
    [
      "P"
    ]
  ],
  "Actions": [
    {
      "Src": 2,
      "Dst": 1,
      "Amount": 1
    },
    {
      "Src": 1,
      "Dst": 3,
      "Amount": 3
    },
    {
      "Src": 2,
      "Dst": 1,
      "Amount": 2
    },
    {
      "Src": 1,
      "Dst": 2,
      "Amount": 1
    }
  ]
}